src/rgw/rgw_rados.cc

```diff
@@ -8463,9 +8463,19 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
     if (r >= 0) {
       ldout(cct, 10) << "RGWRados::" << __func__ << ": got " <<
 	dirent.key.name << "[" << dirent.key.instance << "]" << dendl;
-      m[name] = std::move(dirent);
-      last_entry_visited = &(m[name]);
-      ++count;
+
+      auto [it, success] = m.insert(std::make_pair(name, std::move(dirent)));
+      if (success) {
+	last_entry_visited = &it->second;
+	++count;
+      } else {
+	// NB: this is *never* expected to be executed
```

Comment:
i'm not a fan of dead code. i'd rather we just ignore the `success` flag the same way `m[name] = std::move(dirent);` did

Reply:
I'd prefer an assert to just ignoring, but I know those are controversial.

Reply:
Because if the insert should fail then the last_entry_visited assignment is wonky, which could lead to a runtime error. At least the assert would catch it and explain the failure.

Reply:
i see what you mean. i guess the equivalent of current behavior would be more like this:
```c++
auto [it, inserted] = m.try_emplace(name, std::move(dirent)); // does not move from dirent unless inserted
if (!inserted) {
  it->second = std::move(dirent); // overwrite existing entry
}
last_entry_visited = &it->second;
++count;
```
what do you think?

Reply:
it looks like there's a [map::insert_or_assign()](https://en.cppreference.com/w/cpp/container/map/insert_or_assign) that would work here too

Reply:
You're right, I think this is the direction to go in. Although in an overwrite situation we wouldn't increment count. I'll update the commit and pr. Thanks!

Reply:
this is cool;  I certainly have been concerned about asserts, especially in Ceph historically, as Ceph I think still jumps through hoops to cause asserts to fire (hence abort) in production builds;  but here, it seems like if there were a failure case, it would be recoverable, so not something we would want to terminate for

Reply:
Agreed, @mattbenjamin, especially @cbodley's ideas cover all cases w/o error.